c/examples/connection_retry.c

```diff
@@ -0,0 +1,123 @@
+
+static PolarisContext_t context;
+
+void HandleData(const uint8_t* buffer, size_t size_bytes) {
+  P1_printf("Application received %u bytes.\n", (unsigned)size_bytes);
+}
+
+void HandleSignal(int sig) {
+  signal(sig, SIG_DFL);
+
+  P1_printf("Caught signal %s (%d). Closing Polaris connection.\n",
+            sys_siglist[sig], sig);
+  Polaris_Disconnect(&context);
+}
+
+int main(int argc, const char* argv[]) {
+  if (argc != 3) {
+    P1_fprintf(stderr, "Usage: %s API_KEY UNIQUE_ID\n", argv[0]);
+    return 1;
+  }
+
+  const char* api_key = argv[1];
+  const char* unique_id = argv[2];
+
+  const int MAX_RECONNECTS = 2;
+  int auth_valid = 0;
+  int reconnect_count = 0;
+  while (1) {
+    // Retrieve an access token using the specified API key.
+    if (!auth_valid) {
+      if (Polaris_Init(&context) != POLARIS_SUCCESS) {
+        return 2;
+      }
+
+      P1_printf("Opened Polaris context. Authenticating...\n");
+
+      int ret = Polaris_Authenticate(&context, api_key, unique_id);
+      if (ret == POLARIS_FORBIDDEN) {
+        P1_printf("Authentication rejected. Is your API key valid?\n");
+        return 3;
+      } else if (ret != POLARIS_SUCCESS) {
+        continue;
+      }
+
+      auth_valid = 1;
+    }
+
+    // We now have a valid access token. Connect to the corrections service.
+    //
+    // In the calls below, if the connection times out or the access token is
+    // rejected, try to connect again. If it fails too many times, the access
+    // token may be expired - try reauthenticating.
+    P1_printf("Authenticated. Connecting to Polaris...\n");
+
+    Polaris_SetRTCMCallback(&context, HandleData);
+
+    if (Polaris_Connect(&context) != POLARIS_SUCCESS) {
+      if (++reconnect_count >= MAX_RECONNECTS) {
+        P1_printf(
+            "Max reconnects exceeded. Clearing access token and retrying "
+            "authentication.\n");
+        auth_valid = 0;
+        reconnect_count = 0;
+      }
+      continue;
+    }
+
+    P1_printf("Connected to Polaris...\n");
+
+    // Send a position update to Polaris. Position updates are used to select an
+    // appropriate corrections stream, and should be updated periodically as the
+    // receiver moves.
+    if (Polaris_SendECEFPosition(&context, -2707071.0, -4260565.0, 3885644.0) !=
+        POLARIS_SUCCESS) {
+      Polaris_Disconnect(&context);
+      if (++reconnect_count >= MAX_RECONNECTS) {
+        P1_printf(
+            "Max reconnects exceeded. Clearing access token and retrying "
+            "authentication.\n");
+        auth_valid = 0;
+        reconnect_count = 0;
+      }
+      continue;
+    }
+
+    // Receive incoming RTCM data until the application exits.
+    P1_printf("Sent position. Listening for data...\n");
+
+    signal(SIGINT, HandleSignal);
+    signal(SIGTERM, HandleSignal);
+
+    int ret = Polaris_Run(&context, 30000);
+    if (ret == POLARIS_CONNECTION_CLOSED) {
+      P1_printf("Connection terminated.\n");
+      break;
```

Comment:
I'm confused...why is this a break? What would happen if the server gracefully closes (i.e. during an upgrade of the server)

Reply:
I did this so the program exits if the user hits ctrl-c since `Polaris_Disconnect()` and server disconnect both return `POLARIS_CONNECTION_CLOSED`. I think I'm gonna change that and have Run() return `POLARIS_SUCCESS` if the user calls disconnect. Just requires storing a flag in the context to tell the difference since we'll get the same response from `recv()` either way.